### ISSUE TITLE
Fix broken links in compiler docs

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#73749]
I noticed a bunch of warnings while documenting the compiler:

<details><summary>Warnings</summary>

```
warning: `[exhaustiveness]` cannot be resolved, ignoring it.
 --> src/librustc_mir_build/hair/pattern/_match.rs:7:69
  |
7 | //! (a) the patterns cover every possible constructor for the type [exhaustiveness]
  |                                                                     ^^^^^^^^^^^^^^ cannot be resolved, ignoring
  |
  = note: `#[warn(intra_doc_link_resolution_failure)]` on by default
  = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`

warning: `[usefulness]` cannot be resolved, ignoring it.
 --> src/librustc_mir_build/hair/pattern/_match.rs:8:36
  |
8 | //! (b) each pattern is necessary [usefulness]
  |                                    ^^^^^^^^^^ cannot be resolved, ignoring
  |
  = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`

warning: could not parse code block as Rust code
   --> src/librustc_mir_build/hair/pattern/_match.rs:113:9
    |
113 |   //!        For example, the first pattern above initially gives a stack `[(Some(true), _)]`. If we
    |  _________^
114 | | //!        pop the tuple constructor, we are left with `[Some(true), _]`, and if we then pop the
115 | | //!        `Some` constructor we get `[true, _]`. If we had popped `None` instead, we would get
116 | | //!        nothing back.
...   |
149 | | //!     Note that the OR-patterns are not always used directly in Rust, but are used to derive the
150 | | //!     exhaustive integer matching rules, so they're written here for posterity.
    | |_________________________________________________________________________________^
    |
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: character literal may only contain one codepoint
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: \u{2260}
    = note: error from rustc: unterminated character literal

warning: could not parse code block as Rust code
   --> src/librustc_mir_build/hair/pattern/_match.rs:179:9
    |
179 |   //!           For example, if `P` is:
    |  _________^
180 | | //!           [
181 | | //!               [Some(true), _],
182 | | //!               [None, 0],
...   |
229 | | //!           `U(P, p) := U(P, (r_1, p_2, .., p_n))
230 | | //!                    || U(P, (r_2, p_2, .., p_n))`
    | |____________________________________________________^
    |
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: \u{2203}
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `

warning: 4 warnings emitted
```

</details>

A lot of these look like broken links. It would be great to fix them.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/pull/73748#issuecomment-649908647_

**Comment by lzutao** [comment]
So can we make those warnings into errors ?

**Comment by jyn514** [comment]
We can certainly `deny(intra_doc_resolution_failure)`. The second I'm not sure about, rustc already uses `deny(warnings)` but rustdoc has a global `allow(warnings)` that might be interfering.

**Comment by jyn514** [comment]
@GuillaumeGomez said:

> They're not tracked for the moment until we have the intra-doc links stabilized. I definitely plan to deny this warning as soon as it's stable. :)

So it's definitely good to fix the failures but we should wait on `deny(intra_doc_resolution_failure)` for now.

**Comment by AlphaHot** [comment]
@jyn514  where can I view the source code for `intra_doc_resolution_failure`?

**Comment by jyn514** [comment]
The docs for the lint are [in the rustdoc book](https://doc.rust-lang.org/rustdoc/lints.html#intra_doc_link_resolution_failure). The source code is in https://github.com/rust-lang/rust/blob/ecc6f5683b6c2ad1e92b58011c828cb604e0b5ac/src/librustdoc/passes/collect_intra_doc_links.rs#L908

**Comment by jyn514** [comment]
Closing in favor of https://github.com/rust-lang/rust/issues/80369, I think the intra-doc links have all been fixed.
